### PR TITLE
Fix @require with the same data key and accept all hashable types of variable in @will_generate arguments

### PR DESCRIPTION
**Pull request by penolove**

### CODE REVIEW

File: examples/lifetime_prediction/lifetime_feature_generator.py [hunk 1]
```diff
@@ -17,8 +17,12 @@ def __init__(self, h5py_hdf_path, data_csv_path):
         self.data_csv_path = data_csv_path
 
     @will_generate('memory', 'data_df')
-    def gen_data_df(self):
-        return {'data_df': pd.read_csv(self.data_csv_path, index_col='id')}
+    def gen_data_df(self, args=None):
```
Comment (ianlini) [inline examples/lifetime_prediction/lifetime_feature_generator.py]: should add a @params('nrows') here, and we don't support default value now, so you should give it a value.

File: dagian/data_definition.py [hunk 2]
```diff
@@ -124,6 +124,8 @@ def eval_data_definition(self, args):
                 new_args[key] = arg.eval(args)
             elif isinstance(arg, basestring):
                 new_args[key] = arg.format(**args)
+            elif isinstance(arg, int):
+                new_args[key] = arg
```
Comment (ianlini) [inline dagian/data_definition.py]: just change else-clause to `new_arg[key] = arg`

File: examples/lifetime_prediction/.dagianrc/bundle_config.yml [hunk 3]
```diff
@@ -20,6 +20,9 @@ structure:
       args:
         dividend: weight
         divisor: height
+  unittest:
```
Comment (ianlini) [inline examples/lifetime_prediction/.dagianrc/bundle_config.yml]: don't do this, think of a meaningful word, and why do you need to add a new dataset for this?

File: examples/lifetime_prediction/lifetime_feature_generator.py (hunk 1 above)
Comment (ianlini) [reply examples/lifetime_prediction/lifetime_feature_generator.py]: next step, I will limit the args to only the variables declared in @params, and the default value is maybe the next next ... next step

Comment (ianlini) [reply examples/lifetime_prediction/lifetime_feature_generator.py]: my "default value" here means you should never use `args.get()`. `args=None` is definitely wrong. You will always get this value in the future.

File: dagian/data_definition.py (hunk 2 above)
Comment (ianlini) [reply dagian/data_definition.py]: Sorry, we should check whether `arg` is hashable, so you should put `new_arg[key] = arg` in `elif isinstance(arg, collections.Hashable):`, and change the error message in `else:` to `The values in RequirementDefinition.args can only be Argument or hashable.`

File: examples/lifetime_prediction/tests.py [hunk 4]
```diff
@@ -38,9 +38,9 @@ def test_generate_lifetime_features(self):
         with h5py.File(h5py_hdf_path, "r") as global_data_h5f, \
                 h5py.File(data_bundle_hdf_path, "r") as data_bundle_h5f:
             assert (set(global_data_h5f)
-                    == set(data_definition.json() for data_definition in data_definitions))
+                    > set(data_definition.json() for data_definition in data_definitions))
```
Comment (penolove) [inline examples/lifetime_prediction/tests.py]: set(global_data_h5f) will have two more file than data_definitions from:
```
    @require('is_in_test_set', 'is_in_test_set_1126', random_state=1126)
    @require('is_in_test_set', 'is_in_test_set_5566', random_state=5566)
```

File: dagian/data_definition.py (hunk 2 above)
Comment (ianlini) [reply dagian/data_definition.py]: 你沒改到這個

File: dagian/data_definition.py [hunk 5]
```diff
@@ -125,8 +125,7 @@ def eval_data_definition(self, args):
             elif isinstance(arg, basestring):
                 new_args[key] = arg.format(**args)
             else:
-                raise ValueError(
-                    "The values in RequirementDefinition.args can only be Argument or str.")
+                new_args[key] = arg
```
Comment (ianlini) [inline dagian/data_definition.py]: We should check whether `arg` is hashable, so you should put `new_arg[key] = arg` in `elif isinstance(arg, collections.Hashable):`, and change the error message in `else:` to `The values in RequirementDefinition.args can only be Argument or hashable.`

File: examples/lifetime_prediction/.dagianrc/bundle_config.yml [hunk 6]
```diff
@@ -21,6 +24,8 @@ structure:
         dividend: weight
         divisor: height
 
+
```
Comment (ianlini) [inline examples/lifetime_prediction/.dagianrc/bundle_config.yml]: ??

File: examples/lifetime_prediction/lifetime_feature_generator.py [hunk 7]
```diff
@@ -9,16 +10,19 @@
...
     def __init__(self, h5py_hdf_path, data_csv_path):
         super(LifetimeFeatureGenerator, self).__init__(
             h5py_hdf_path=h5py_hdf_path)
         self.data_csv_path = data_csv_path
 
     @will_generate('memory', 'data_df')
-    def gen_data_df(self):
-        return {'data_df': pd.read_csv(self.data_csv_path, index_col='id')}
+    def gen_data_df(self, args=None):
+        if args is None:
```
Comment (ianlini) [inline examples/lifetime_prediction/lifetime_feature_generator.py]: It's impossible to be None

File: examples/lifetime_prediction/lifetime_feature_generator.py (hunk 1 above)
Comment (ianlini) [reply examples/lifetime_prediction/lifetime_feature_generator.py]: You don't understand...

Comment (ianlini) [reply examples/lifetime_prediction/lifetime_feature_generator.py]: You should never use args.get(). args=None is definitely wrong. You will always get this value in the future.
You are still using them.

Comment (ianlini) [reply examples/lifetime_prediction/lifetime_feature_generator.py]: and you didn't add `@params`

File: examples/lifetime_prediction/tests.py (hunk 4 above)
Comment (ianlini) [reply examples/lifetime_prediction/tests.py]: then you should add them

File: examples/lifetime_prediction/tests.py [hunk 8]
```diff
@@ -37,10 +37,17 @@ def test_generate_lifetime_features(self):
...
         with h5py.File(h5py_hdf_path, "r") as global_data_h5f, \
                 h5py.File(data_bundle_hdf_path, "r") as data_bundle_h5f:
-            assert (set(global_data_h5f)
-                    == set(data_definition.json() for data_definition in data_definitions))
+            data_definitions_json_set = set(
+                data_definition.json() for data_definition in data_definitions)
+
+            # additional two data_definitions from test_masks
+            additional_data_definition = {
+                '{"key": "is_in_test_set", "args": {"random_state": 1126}}',
```
Comment (ianlini) [inline examples/lifetime_prediction/tests.py]: You should use `DataDefinition` to define these and use `.json()` to transform it to json.

File: examples/lifetime_prediction/tests.py [hunk 9]
```diff
@@ -37,10 +37,17 @@ def test_generate_lifetime_features(self):
...
+                '{"key": "is_in_test_set", "args": {"random_state": 5566}}'
```
Comment (ianlini) [inline examples/lifetime_prediction/tests.py]: comma

File: examples/lifetime_prediction/tests.py [hunk 10]
```diff
@@ -37,10 +37,17 @@ def test_generate_lifetime_features(self):
         data_definitions = get_data_definitions_from_structure(bundle_config['structure'])
         with h5py.File(h5py_hdf_path, "r") as global_data_h5f, \
                 h5py.File(data_bundle_hdf_path, "r") as data_bundle_h5f:
-            assert (set(global_data_h5f)
-                    == set(data_definition.json() for data_definition in data_definitions))
+            data_definitions_json_set = set(
```
Comment (ianlini) [inline examples/lifetime_prediction/tests.py]: `data_definitions_json_set` -> `cache_data_def_json_set`

File: examples/lifetime_prediction/tests.py [hunk 11]
```diff
@@ -37,10 +37,17 @@ def test_generate_lifetime_features(self):
...
+                data_definition.json() for data_definition in data_definitions)
+
+            # additional two data_definitions from test_masks
+            additional_data_definition = {
+                '{"key": "is_in_test_set", "args": {"random_state": 1126}}',
+                '{"key": "is_in_test_set", "args": {"random_state": 5566}}'
+            }
+            assert (set(global_data_h5f) == data_definitions_json_set | additional_data_definition)
```
Comment (ianlini) [inline examples/lifetime_prediction/tests.py]: `global_data_h5f` -> `cache_data_h5f`

File: examples/lifetime_prediction/tests.py [hunk 12]
```diff
@@ -37,10 +37,17 @@ def test_generate_lifetime_features(self):
         data_definitions = get_data_definitions_from_structure(bundle_config['structure'])
         with h5py.File(h5py_hdf_path, "r") as global_data_h5f, \
                 h5py.File(data_bundle_hdf_path, "r") as data_bundle_h5f:
-            assert (set(global_data_h5f)
-                    == set(data_definition.json() for data_definition in data_definitions))
+            data_definitions_json_set = set(
+                data_definition.json() for data_definition in data_definitions)
+
+            # additional two data_definitions from test_masks
+            additional_data_definition = {
```
Comment (ianlini) [inline examples/lifetime_prediction/tests.py]: directly add it to cache_data_def_json_set

**Comment by ianlini** [comment]
Thank you @penolove. You are the first external contributor. Congratulations! :tada:

**Pull request merged**